src/jdk.jdwp.agent/unix/native/libjdwp/exec_md.c

```diff
@@ -44,6 +48,102 @@ static char *skipNonWhitespace(char *p) {
     return p;
 }
 
+#if defined(_AIX)
+  /* AIX does not understand '/proc/self' - it requires the real process ID */
+  #define FD_DIR aix_fd_dir
+  #define DIR DIR64
+  #define dirent dirent64
+  #define opendir opendir64
+  #define readdir readdir64
+  #define closedir closedir64
+#elif defined(_ALLBSD_SOURCE)
+  #define FD_DIR "/dev/fd"
+#else
+  #define FD_DIR "/proc/self/fd"
+#endif
+
+// Closes every file descriptor that is listed as a directory
+// entry in "/proc/self/fd" (or its equivalent). Standard
+// input/output/error file descriptors will not be closed
+// by this function. This function returns 0 on failure
+// and 1 on success.
+int
+closeDescriptors(void)
+{
+    DIR *dp;
+    struct dirent *dirp;
+    /* leave out standard input/output/error descriptors */
+    int from_fd = STDERR_FILENO + 1;
+
+    /* We're trying to close all file descriptors, but opendir() might
+     * itself be implemented using a file descriptor, and we certainly
+     * don't want to close that while it's in use.  We assume that if
+     * opendir() is implemented using a file descriptor, then it uses
+     * the lowest numbered file descriptor, just like open().  So
+     * before calling opendir(), we close a couple explicitly, so that
+     * opendir() can then use these lowest numbered closed file
+     * descriptors afresh.   */
+
+    close(from_fd);          /* for possible use by opendir() */
+    close(from_fd + 1);      /* another one for good luck */
+    from_fd += 2; /* leave out the 2 we just closed, which the opendir() may use */
+
+#if defined(_AIX)
+    /* set FD_DIR for AIX which does not understand '/proc/self' - it
+     * requires the real process ID */
+    char aix_fd_dir[32];     /* the pid has at most 19 digits */
+    snprintf(aix_fd_dir, 32, "/proc/%d/fd", getpid());
+#endif
+
+    if ((dp = opendir(FD_DIR)) == NULL) {
+        LOG_MISC(("warning: failed to open dir %s while determining"
+                  " file descriptors to close for process %d", FD_DIR, getpid()));
+        return 0; // failure
+    }
+
+    while ((dirp = readdir(dp)) != NULL) {
+        int fd;
+        if (isdigit(dirp->d_name[0]) &&
+            (fd = strtol(dirp->d_name, NULL, 10)) >= from_fd) {
+            (void)close(fd);
```

Comment:
I'd really, really like to check `close()` for errors and report any.

Reply:
If we are going to check close for errors then we will have to know for certain we are only trying to close open fd's, and we will have to deal with EINTR. I would think this is a "best effort" to close open FD's and not something we have to care about in detail.

Reply:
Looking at man close, EIO gets returned if:
> [EIO]              A previously-uncommitted write(2) encountered an input/output error.

If I understand this `launch=<cmd>` option correctly for the jdwp agent, then the arbitrary command gets launched very early in the lifecycle of the JVM. So in theory, I think there shouldn't be too many file descriptors that the JVM has opened so far and the chances of it having started writing to some of those opened file descriptors (other than standard out/err) perhaps is slim? But that's just a guess. Probably not a strong one too, given that if the JVM was launched with any other agent along with jdwp, then it's unknown if the other agent would have opened any file descriptors for write. 

My understanding of these close() calls after fork() in this code, matches David's - I think it's a best effort attempt. So perhaps ignoring such failures, like we do now, is OK?

Reply:
Ignoring errors just doesn't seem like a good strategy.

This fix right here that we are doing doesn't have to deal with making this code more robust, as long as we follow up in another issue. For that, however, I filed JDK-8324979, and that frees us up to move on with this fix.